File: app/Models/CheckIn.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

class CheckIn extends Model
{
    use HasFactory;

    protected $fillable = [
        'schedule_id', 'user_id', 'store_id', 'latitude', 'longitude',
        'photo_path', 'is_verified', 'distance_from_store', 'checked_in_at',
    ];

    protected function casts(): array
    {
        return [
            'latitude' => 'decimal:8',
            'longitude' => 'decimal:8',
            'distance_from_store' => 'decimal:2',
            'is_verified' => 'boolean',
            'checked_in_at' => 'datetime',
        ];
    }

    public function schedule()
    {
        return $this->belongsTo(Schedule::class);
    }

    public function user()
    {
        return $this->belongsTo(User::class);
    }

    public function store()
    {
        return $this->belongsTo(Store::class);
    }

    /**
     * Calculate distance between two GPS coordinates using the Haversine formula.
     * Returns distance in meters.
     */
    public static function calculateDistance($lat1, $lon1, $lat2, $lon2): float
    {
        $earthRadius = 6371000; // meters

        $dLat = deg2rad($lat2 - $lat1);
        $dLon = deg2rad($lon2 - $lon1);

        $a = sin($dLat / 2) * sin($dLat / 2) +
             cos(deg2rad($lat1)) * cos(deg2rad($lat2)) *
             sin($dLon / 2) * sin($dLon / 2);

        $c = 2 * atan2(sqrt($a), sqrt(1 - $a));

        return $earthRadius * $c;
    }
}

File: mobile/lib/models/check_in.dart
class CheckInModel {
  final int id;
  final Map<String, dynamic> store;
  final double latitude;
  final double longitude;
  final String photoUrl;
  final bool isVerified;
  final double distanceFromStore;
  final String checkedInAt;

  CheckInModel({
    required this.id,
    required this.store,
    required this.latitude,
    required this.longitude,
    required this.photoUrl,
    required this.isVerified,
    required this.distanceFromStore,
    required this.checkedInAt,
  });

  factory CheckInModel.fromJson(Map<String, dynamic> json) {
    return CheckInModel(
      id: json['id'],
      store: json['store'],
      latitude: double.parse(json['latitude'].toString()),
      longitude: double.parse(json['longitude'].toString()),
      photoUrl: json['photo_url'],
      isVerified: json['is_verified'],
      distanceFromStore: double.parse(json['distance_from_store'].toString()),
      checkedInAt: json['checked_in_at'],
    );
  }
}
